File: python3/longest_substring_without_repeating_characters.py
"""
Space   : O(n)
Time    : O(n)
"""

class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        ans = 0
        start = 0
        d = {}
        
        for i, x in enumerate(s):
            if x in d and start <= d[x]:
                start = d[x]+1
            else:
                ans = max(ans, i - start+1)
            d[x] = i
    
        return ans

File: golang/longest_substring_without_repeating_characters.go
package main

// Space	: O(n)
// Time		: O(n)

func lengthOfLongestSubstring(s string) int {
	start, ans := 0, 0
	d := map[byte]int{}

	for i := 0; i < len(s); i++ {
		if _, ok := d[s[i]]; ok && start <= d[s[i]] {
			start = d[s[i]] + 1
		} else {
			ans = Max(ans, i-start+1)
		}
		d[s[i]] = i
	}

	return ans
}
